Event start

StoyanGoranov opened the pull request:
Closes: https://github.com/Not-Endurance/not-timing-system/issues/282

StoyanGoranov committed:
use AthleteCategory instead of string for Athlete Category field

at Setup Domain (Entity Model, FormModel and Form Input)

https://github.com/Not-Endurance/not-timing-system/issues/282

StoyanGoranov committed:
add OverrideMaxSpeed field in ContestantForm at SetupDomain

https://github.com/Not-Endurance/not-timing-system/issues/282

StoyanGoranov committed:
separate Competition Type and Competition Ruleset properties

https://github.com/Not-Endurance/not-timing-system/issues/282

StoyanGoranov committed:
add minor changes to Competition, Contestant Forms

https://github.com/Not-Endurance/not-timing-system/issues/282

StoyanGoranov committed:
add Ranking repository

https://github.com/Not-Endurance/not-timing-system/issues/282

StoyanGoranov committed:
implement CreateParticipations and CreateRankings method

https://github.com/Not-Endurance/not-timing-system/issues/282

StoyanGoranov committed:
Merge branch 'develop' of https://github.com/Not-Endurance/not-timing-system into event-start

StoyanGoranov committed:
add default CompetitionType and use it at EmsImport

maintain EmsImportBehind compatibility

https://github.com/Not-Endurance/not-timing-system/issues/282

StoyanGoranov committed:
add CreateEvent Validation at DashboardBehind

https://github.com/Not-Endurance/not-timing-system/issues/282

StoyanGoranov committed:
implement EnduranceEventStatePanel and use it at NavMenu

https://github.com/Not-Endurance/not-timing-system/issues/282

achobanov commented on src-v2/Judge/NTS.Judge.Blazor/Pages/Setup/Athletes/AthleteForm.razor:
```diff
@@ -13,25 +14,27 @@
-    <MudTextField @bind-Value="_model.Club" Label="Club" Variant="Variant.Text" Margin="Margin.Dense" />
-    <MudTextField @bind-Value="_model.Category" Label="Category" Variant="Variant.Text" Margin="Margin.Dense" />
+    <MudTextField @bind-Value="_model.Club" Label="@Localizer.Get("Club")" Variant="Variant.Text" Margin="Margin.Dense" />
+    <NotSelect @ref="_categoryField" T="AthleteCategory" Items="_categories" @bind-Value="_model.Category" Label="@Localizer.Get("Category")"/>
```
Is Localizer necessary in `NotSelect.Label`?

achobanov commented on src-v2/Judge/NTS.Judge.Blazor/Shared/EnduranceEventStatePanel.razor:
```diff
@@ -0,0 +1,20 @@
+@inherits NotComponent
+
+<MudStack Class="ma-6">
+    <MudButton Color="Color.Primary" Variant="Variant.Filled" OnClick="Start">@Localizer.Get("Start Endurance Event")</MudButton>
```
Use `NotButtonPrimary` instead

achobanov commented on src-v2/Judge/NTS.Judge/Adapters/Behinds/DashboardBehind.cs:
```diff
@@ -36,7 +44,20 @@ async Task SafeStart()
             throw new DomainException("Cannot start - event is not configured");
         }
         await CreateEvent(setupEvent);
+        if (!setupEvent.Officials.Any())
+        {
+            throw new DomainException("Cannot start - officials aren't configured");
+        }
```
Are officials required?

achobanov commented on src-v2/Judge/NTS.Judge/Adapters/Behinds/DashboardBehind.cs:
```diff
@@ -36,7 +44,20 @@ async Task SafeStart()
+            throw new DomainException("Cannot start - officials aren't configured");
+        }
         await CreateOfficials(setupEvent.Officials);
+        if (!setupEvent.Competitions.ElementAt(0).Phases.Any())
+        {
+            throw new DomainException("Cannot start - phases aren't configured");
+        }
+        if (!setupEvent.Competitions.ElementAt(0).Contestants.Any())
+        {
+            throw new DomainException("Cannot start - contestants aren't configured");
+        }
```
This isn't a very good validation:
1. Competitions may be empty, which will throw unhandled exception
2. You are only validating the first competition, which means you can start with second competition without phases
3. The message isn't very descriptive, because the user may have configured phases, but omitted one of the competitions.

achobanov commented on src-v2/Judge/NTS.Judge/Adapters/Behinds/DashboardBehind.cs:
```diff
@@ -63,6 +84,68 @@ async Task CreateOfficials(IEnumerable<Domain.Setup.Entities.Official> setupOffi
+            var setupPhases = competition.Phases;
+            var competitionDistance = 0m;
+            var phases = new List<Phase>();
+            var categoryEntriesPairs = new Dictionary<AthleteCategory, List<RankingEntry>>
```
This isn't a very good dictionary name:
1. What entries
2. What pairs

A good naming for dictionaries is $"{values}By{Category}", aka: rankingEntriesByCategory.

achobanov commented on src-v2/Judge/NTS.Judge/Adapters/Behinds/DashboardBehind.cs:
```diff
@@ -63,6 +84,68 @@ async Task CreateOfficials(IEnumerable<Domain.Setup.Entities.Official> setupOffi
+            };
+            foreach(var phase in setupPhases)
+            {
+                phases.Add(new Phase(phase.Loop!.Distance, phase.Recovery, phase.Rest, competition.Ruleset, setupPhases.Last() == phase, competition.CriRecovery));
```
nested method invocations should be avoided as it's hard to tell which parameter belongs to which method and it leads to inferior debugging experience. Instead create the `Phase` as variable.

achobanov commented on src-v2/Judge/NTS.Judge/Adapters/Behinds/DashboardBehind.cs:
```diff
@@ -63,6 +84,68 @@ async Task CreateOfficials(IEnumerable<Domain.Setup.Entities.Official> setupOffi
+                var maxSpeed = (double?)null;
+                var minSpeed = (double?)null;
+                //consider adding magic ints to StaticOptions or constants
+                if(competition.Type == CompetitionType.Qualification)
+                {
+                    minSpeed = 10;
+                    maxSpeed = 16;
+                }
+                if(combination.Athlete.Category == AthleteCategory.Children)
+                {
+                    minSpeed = 8;
+                    maxSpeed = 12;
+                }
+                maxSpeed = contestant.MaxSpeedOverride != null ? contestant.MaxSpeedOverride : maxSpeed;
```
This is business logic and thus should be reflected in the Domain Model. I think you can safely internalize it in the Tandem ctor

achobanov commented on src-v2/Judge/NTS.Judge/Adapters/Behinds/DashboardBehind.cs:
```diff
@@ -63,6 +84,68 @@ async Task CreateOfficials(IEnumerable<Domain.Setup.Entities.Official> setupOffi
+                var participation = new Participation(competition.Name, competition.Ruleset, tandem, phases);
+                await _participationRepository.Create(participation);
+                var rankingEntry = new RankingEntry(participation, !contestant.IsUnranked);
+                foreach(var pair in categoryEntriesPairs)
+                {
+                    if(pair.Key == combination.Athlete.Category)
+                    {
+                        pair.Value.Add(rankingEntry);
+                    }
+                }
```
Why foreach here when you can simply access by key?

achobanov commented on src-v2/Judge/NTS.Judge/Adapters/Behinds/DashboardBehind.cs:
```diff
@@ -63,6 +84,68 @@ async Task CreateOfficials(IEnumerable<Domain.Setup.Entities.Official> setupOffi
         }
     }
 
+    async Task CreateParticipations(Domain.Setup.Entities.Event setupEvent) 
```
This method name is misleading as it creates Rankinigs as well.

achobanov commented on src-v2/Judge/NTS.Judge/Adapters/Behinds/DashboardBehind.cs:
```diff
@@ -63,6 +84,68 @@ async Task CreateOfficials(IEnumerable<Domain.Setup.Entities.Official> setupOffi
+                maxSpeed = contestant.MaxSpeedOverride != null ? contestant.MaxSpeedOverride : maxSpeed;
+                var tandem = new Tandem(combination.Number, combination.Athlete.Person, combination.Horse.Name, competitionDistance, combination.Athlete.Country, combination.Athlete.Club, minSpeed, maxSpeed);
+                var participation = new Participation(competition.Name, competition.Ruleset, tandem, phases);
+                await _participationRepository.Create(participation);
```
It's good idea to separate the domain creation logic and storage. Maybe something like StartFactory, DashboardFactory or CoreFactory class, which contains: `CreateParticipation`, `CreateRanking`, `CreateOfficial` etc

achobanov commented on src-v2/NTS.Domain.Setup/Entities/Competition.cs:
```diff
@@ -24,36 +25,40 @@ public static Competition Update(
         if (type == default)
         {
-            throw new DomainException(nameof(type), "Competition Type is required");
+            //CompetitionType defualt is enabled to ensure EmsImportBehind function correctly
+            //throw new DomainException(nameof(type), "Competition Type is required");
```
Is it not possible to avoid disabling validation? Why not just pass the argument?

StoyanGoranov replied:
Currently, the EmsImportBehind calls `Competition.Create` with a default value for `type` this will always cause the EmsImport to throw an exception.

achobanov commented on src-v2/NTS.Domain/Enums/CompetitionType.cs:
```diff
@@ -0,0 +1,12 @@
+
+public enum CompetitionType
+{
+    TypeNotSet = 0,
```
We don't provide default values in enums

achobanov commented on src-v2/NTS.Persistance/Adapters/RankingRepository.cs:
```diff
@@ -0,0 +1,8 @@
+
+namespace NTS.Persistence.Adapters;
+
+public class RankingRepository(IStore<CoreState> store) : SetRepository<Ranking, CoreState>(store)
+{
+}
```
As it is currently there are two `IRepository<Ranking>` implementations. Rename ClassificationRepository instead of creating a new RankingRepository.

StoyanGoranov committed:
rename ClassificationRepository to Ranking

https://github.com/Not-Endurance/not-timing-system/issues/282

StoyanGoranov committed:
clean code and improve names at DashboardBehind

https://github.com/Not-Endurance/not-timing-system/issues/282

StoyanGoranov committed:
move min and max Tandem Speed logic at Domain Tandem.cs

https://github.com/Not-Endurance/not-timing-system/issues/282

StoyanGoranov committed:
adhere to Not code conventions

https://github.com/Not-Endurance/not-timing-system/issues/282

StoyanGoranov committed:
modify Tandem ctor call at ACL\Participations

https://github.com/Not-Endurance/not-timing-system/issues/282

StoyanGoranov committed:
enable Competition type validation, complete EmsImportBehind MapRuleset
logic

https://github.com/Not-Endurance/not-timing-system/issues/282

StoyanGoranov committed:
add CoreFactory.cs to create Core Domain Entities from Setup Entities

moving entity creation logic and validation from DashboardBehind into
CoreFactory; using CoreFactory in DashboardBehind

https://github.com/Not-Endurance/not-timing-system/issues/282

achobanov commented on src-v2/Judge/NTS.Judge/Adapters/Behinds/DashboardBehind.cs:
```diff
@@ -41,110 +40,52 @@ async Task SafeStart()
-                maxSpeed = contestant.MaxSpeedOverride != null ? contestant.MaxSpeedOverride : maxSpeed;
-                var tandem = new Tandem(combination.Number, combination.Athlete.Person, combination.Horse.Name, competitionDistance, combination.Athlete.Country, combination.Athlete.Club, minSpeed, maxSpeed);
-                var participation = new Participation(competition.Name, competition.Ruleset, tandem, phases);
+            var participationsAndRankingEntries = CoreFactory.CreateParticipationAndRankingEntries(competition);
+            var participations = participationsAndRankingEntries.Participations;
+            var rankingEntriesByCategory = participationsAndRankingEntries.RankingEntriesByCategory;
```
can do 
```
var (participations, entriesByCategory) = CoreFactory.Create....
```

achobanov commented on src-v2/NTS.Domain.Core/Aggregates/Participations/Tandem.cs:
```diff
@@ -5,6 +5,11 @@ namespace NTS.Domain.Core.Aggregates.Participations;
 // TODO: probably shoudl be a record
 public class Tandem : DomainEntity
 {
+    const double CHILDREN_MIN_SPEED = 8;
+    const double CHILDREN_MAX_SPEED = 12;
+    const double QUALIFICATION_MIN_SPEED = 10;
+    const double QUALIFICATION_MAX_SPEED = 16;
```
As discussed this is a Setup logic.

StoyanGoranov committed:
imrpove assignment of tuple method results

https://github.com/Not-Endurance/not-timing-system/issues/282

StoyanGoranov committed:
add TODO comment at Tandem for DependencyInversion

https://github.com/Not-Endurance/not-timing-system/issues/264

StoyanGoranov committed:
introduce Speed limit logic at Competition Add(Contestant) method

https://github.com/Not-Endurance/not-timing-system/issues/282

StoyanGoranov committed:
fix ICrumbNavigator CanNavigateBack method return type

was void, set to bool

https://github.com/Not-Endurance/not-timing-system/issues/282

StoyanGoranov committed:
clean code, set internal accessibility of SetSpeedLimits method, improve
TODO comments

https://github.com/Not-Endurance/not-timing-system/issues/282
